### ISSUE TITLE
External disk data loss due to Docker entrypoint.sh cruft

### ISSUE THREAD

**Issue by geldot** [ClickHouse/ClickHouse#73949]
This issue is observed under Clickhouse 24.12 deployed with Docker.

We have external disks configured and this has worked fine for several years.

Recently, one of the cold storage disks was removed during maintenance. Our expectation was that Clickhouse would mark the disk ``is_broken`` until the disk could be reconnected.

### Observed behavor
However, it appears that ``entrypoint.sh`` creates empty directories when a disk mount point is missing. The newly created directory ends up in the container's root filesystem, which is ephemeral, typically either an anonymous volume or ``tmpfs``. Hence data can be written to the disk now but that data will disappear silently after each reboot. This behavior is nonsensical, undocumented and generally the exact opposite of what you want when a disk in the configured ``<disks>`` section is missing for any reason.

### Expected behavor
``entrypoint.sh`` should not re-create, chown, or otherwise manipulate disk entries, and should not create directories if they are missing. If the disk's path is missing, Clickhouse will then correctly mark the disk as broken until it is plugged back in.

### Background
The code in ``entrypoint.sh`` parses the user's configuration and arbitrarily creates directories and changes their permissions. The behavior is inherently wrong for external disk paths (``DISKS_PATHS``), and probably wrong for metadata paths too (``DISKS_METADATA_PATHS``) and also questionable for the other directory types -- does anyone remember why this code exists?

The harmful code:

```sh
# There could be many disks declared in config
readarray -t DISKS_PATHS < <(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key='storage_configuration.disks.*.path' || true)
readarray -t DISKS_METADATA_PATHS < <(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key='storage_configuration.disks.*.metadata_path' || true)

CLICKHOUSE_USER="${CLICKHOUSE_USER:-default}"
CLICKHOUSE_PASSWORD_FILE="${CLICKHOUSE_PASSWORD_FILE:-}"
if [[ -n "${CLICKHOUSE_PASSWORD_FILE}" && -f "${CLICKHOUSE_PASSWORD_FILE}" ]]; then
  CLICKHOUSE_PASSWORD="$(cat "${CLICKHOUSE_PASSWORD_FILE}")"
fi
CLICKHOUSE_PASSWORD="${CLICKHOUSE_PASSWORD:-}"
CLICKHOUSE_DB="${CLICKHOUSE_DB:-}"
CLICKHOUSE_ACCESS_MANAGEMENT="${CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT:-0}"

function create_directory_and_do_chown() {
    local dir=$1
    # check if variable not empty
    [ -z "$dir" ] && return
    # ensure directories exist
    if [ "$DO_CHOWN" = "1" ]; then
        mkdir=( mkdir )
    else
        # if DO_CHOWN=0 it means that the system does not map root user to "admin" permissions
        # it mainly happens on NFS mounts where root==nobody for security reasons
        # thus mkdir MUST run with user id/gid and not from nobody that has zero permissions
        mkdir=( clickhouse su "${USER}:${GROUP}" mkdir )
    fi
    if ! "${mkdir[@]}" -p "$dir"; then
        echo "Couldn't create necessary directory: $dir"
        exit 1
    fi

    if [ "$DO_CHOWN" = "1" ]; then
        # ensure proper directories permissions
        # but skip it for if directory already has proper premissions, cause recursive chown may be slow
        if [ "$(stat -c %u "$dir")" != "$USER" ] || [ "$(stat -c %g "$dir")" != "$GROUP" ]; then
            chown -R "$USER:$GROUP" "$dir"
        fi
    fi
}

create_directory_and_do_chown "$DATA_DIR"

# Change working directory to $DATA_DIR in case there're paths relative to $DATA_DIR, also avoids running
# clickhouse-server at root directory.
cd "$DATA_DIR"

for dir in "$ERROR_LOG_DIR" \
  "$LOG_DIR" \
  "$TMP_DIR" \
  "$USER_PATH" \
  "$FORMAT_SCHEMA_PATH" \
  "${DISKS_PATHS[@]}" \
  "${DISKS_METADATA_PATHS[@]}"
do
    create_directory_and_do_chown "$dir"
done
```

**Comment by geldot** [comment]
The code that creates empty directories in place of missing disks was added in #39121 / #17717 cc. @nikitamikhaylov

**Comment by alexey-milovidov** [comment]
@geldot Thanks for the report. I also don't like the cruft in our Docker image - in my opinion, it should not contain any scripts, just the ClickHouse binary. The fact is that ClickHouse does not need Docker, and Docker does not provide any advantages for ClickHouse; it is needed only when some infrastructure requires Docker. Otherwise, it is much simpler not to use Docker.

Please remove this craft and submit a pull request.

**Comment by Adelagric** [comment]
I'd like to pick this up.

I confirmed the behavior is still present on master: `manage_clickhouse_directories` in `docker/server/entrypoint.sh` runs an unconditional `mkdir -p` (+ `chown -R`) over every `storage_configuration.disks.*.path` and `.metadata_path`, so a missing external mount silently becomes an empty directory on the container's ephemeral filesystem.

Two source-level observations that shape the fix:

1. The server does not need the entrypoint to create these directories: `DiskLocal::setup()` already does `fs::create_directories(disk_path)` at disk startup (as the `clickhouse` user), and startup marks the disk broken / fails loudly when that's not possible.
2. Because the server-side creation runs as `clickhouse` rather than root, ordinary Unix permissions already distinguish the two cases: a legitimate first-boot path under the (chowned) data volume gets created by the server itself, while a missing mount under a root-owned parent (e.g. `/mnt/...`) fails loudly instead of being silently fabricated. The entrypoint's root `mkdir -p` is precisely what defeats that protection.

Proposed change, kept minimal in the spirit of reducing entrypoint cruft: for `DISKS_PATHS` and `DISKS_METADATA_PATHS` only, stop creating missing directories, and keep only the useful half of #39121 — chown of directories that do exist (freshly mounted root-owned disks). Default paths (data dir, logs, tmp, user_files, format_schemas) keep today's behavior, so the standard first boot is unaffected.

Before opening the PR I'm running a behavior matrix on the official image (missing mount with and without `skip_access_check`, standard first boot, root-owned mounted disk, disk path under the data dir, cache disks, object-storage `metadata_path`) and will post the results here. Two scope limits I already know about: with `CLICKHOUSE_RUN_AS_ROOT=1` the server recreates the path itself, so this fix doesn't cover that mode; and a missing mount whose parent is writable by `clickhouse` will still be recreated by the server — fully closing that class would need a server-side guard (e.g. a per-disk marker file, similar in spirit to `PG_VERSION` checks), which I'd leave as a follow-up discussion.

@geldot does this match what you had in mind? Happy to adjust before opening the PR.

**Comment by Adelagric** [comment]
Behavior matrix done — results and the fix are in #119009, verified on `clickhouse/clickhouse-server:26.8.2.7`.

Summary: the silent-loss repro (100k rows → `count() = 0` after a container recreate, zero errors, disk reported healthy) is confirmed on the current image. With the fix, a missing mount under a root-owned parent fails loudly at startup by default (`Disk cold is marked as broken during startup ... Permission denied`, then `PATH_ACCESS_DENIED`), and with `skip_access_check=true` on the disk you get exactly the behavior described in this issue: server up, `system.disks.is_broken = 1`, inserts rejected loudly until the disk is reconnected. Standard first boot, root-owned mounted disks (chown preserved), disks under the data dir, and s3/cache disks with not-yet-existing `metadata_path`/cache paths are all unaffected.

**Comment by geldot** [comment]
Thanks for revisiting this @Adelagric. I strongly concur with @alexey-milovidov that we should just remove the cruft, ideally minimizing or deleting the whole `entrypoint.sh` script. Adding more LoC is directionally wrong.

Almost everything the shell script does - creating directories for mount points, changing permissions recursively, etc. is actively harmful.

**Comment by Adelagric** [comment]
Agreed — I've reworked #119009 to be pure deletion: the two `readarray` extractions and the disk entries in the mkdir/chown loop are gone, no new logic added (net −9 LoC). Re-verified the behavior matrix with that variant: the missing-mount case still fails loudly at startup instead of silently writing into the container filesystem, and standard first boot, disks under the data dir, and s3/cache metadata paths are all unchanged.

One additional behavior change to be aware of, noted in the PR: a mounted but root-owned disk is no longer recursively chowned by the entrypoint. The server then comes up with that disk read-only and inserts fail loudly (`TABLE_IS_PERMANENTLY_READ_ONLY`) until the volume's ownership is fixed once by the operator — consistent with the no-magic direction, and loud rather than silent.

Minimizing or deleting the whole `entrypoint.sh` goes beyond this issue's scope and would need a maintainer decision on the supported Docker UX (`CLICKHOUSE_DB`, initdb scripts, user setup) — I'd keep that as a separate discussion.

**Comment by geldot** [comment]
OK. Sounds good as a data-loss hotfix. The remainder of the script is still trash and will need future followup. Review in PR.

(Wish we had done this earlier. It's easy to forget others remain impacted once you apply a local workaround.)
